// @aws-sdk/lib-dynamodb/dist-cjs/baseCommand/DynamoDBDocumentClientCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoDBDocumentClientCommand = void 0;
const smithy_client_1 = require("@smithy/smithy-client");
const utils_1 = require("../commands/utils");
class DynamoDBDocumentClientCommand extends smithy_client_1.Command {
    addMarshallingMiddleware(configuration) {
        const { marshallOptions, unmarshallOptions } = configuration.translateConfig || {};
        this.clientCommand.middlewareStack.addRelativeTo((next, context) => async (args) => {
            args.input = (0, utils_1.marshallInput)(this.input, this.inputKeyNodes, marshallOptions);
            context.dynamoDbDocumentClientOptions =
                context.dynamoDbDocumentClientOptions || DynamoDBDocumentClientCommand.defaultLogFilterOverrides;
            const input = args.input;
            context.dynamoDbDocumentClientOptions.overrideInputFilterSensitiveLog = () => {
                var _a;
                return (_a = context.inputFilterSensitiveLog) === null || _a === void 0 ? void 0 : _a.call(context, input);
            };
            return next(args);
        }, {
            name: "DocumentMarshall",
            relation: "before",
            toMiddleware: "serializerMiddleware",
            override: true,
        });
        this.clientCommand.middlewareStack.addRelativeTo((next, context) => async (args) => {
            const deserialized = await next(args);
            const output = deserialized.output;
            context.dynamoDbDocumentClientOptions =
                context.dynamoDbDocumentClientOptions || DynamoDBDocumentClientCommand.defaultLogFilterOverrides;
            context.dynamoDbDocumentClientOptions.overrideOutputFilterSensitiveLog = () => {
                var _a;
                return (_a = context.outputFilterSensitiveLog) === null || _a === void 0 ? void 0 : _a.call(context, output);
            };
            deserialized.output = (0, utils_1.unmarshallOutput)(deserialized.output, this.outputKeyNodes, unmarshallOptions);
            return deserialized;
        }, {
            name: "DocumentUnmarshall",
            relation: "before",
            toMiddleware: "deserializerMiddleware",
            override: true,
        });
    }
}
exports.DynamoDBDocumentClientCommand = DynamoDBDocumentClientCommand;
DynamoDBDocumentClientCommand.defaultLogFilterOverrides = {
    overrideInputFilterSensitiveLog(...args) { },
    overrideOutputFilterSensitiveLog(...args) { },
};
